tm1637ctl.1
.\"-
.\"Copyright (c) 2021, Alexander Mishin
.\"All rights reserved.
.\"
.\"Redistribution and use in source and binary forms, with or without
.\"modification, are permitted provided that the following conditions are met:
.\"
.\"* Redistributions of source code must retain the above copyright notice, this
.\"  list of conditions and the following disclaimer.
.\"
.\"* Redistributions in binary form must reproduce the above copyright notice,
.\"  this list of conditions and the following disclaimer in the documentation
.\"  and/or other materials provided with the distribution.
.\"
.\"THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\"AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\"IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\"DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
.\"FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\"DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\"SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
.\"CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\"OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\"OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.Dd Jun 28, 2021
.Dt TM1637CTL 1
.Os
.Sh NAME
.Nm tm1637ctl
.Nd tm1637 display control program.
.Sh SYNOPSIS
.Nm
.Op Fl t Ar 0|1
.Op Fl l Ar level
.Op Fl d Ar device
.Sh OPTIONS
The
.Nm
supports the following options:
.Bl -tag -width indent
.It Fl d , Ic --device Ar device
Specify a
.Xr tm1637 4
or
.Xr tm1637d 8
character device. The compiled-in default is
.Pa /dev/tm1637 Ns ;
.It Fl l , Ic --brightness Ar level
Set the display brightness level as an integer 0..7. The display will not
turn on if it is off, but the brightness will be remembered until the next
time it is turned on; 
.It Fl t , Ic --turn Ar 0|1
Turn the display off or on. When it turns on it lights with brightness
previously remembered.
.Sh DESCRIPTION
The
.Nm
is a simple utility to set brightness level and turn on and off the display.
It can be used for manage the display by other program or script (as
.Xr cron 8
job or
.Xr bh1750-daemon 8 Ns ).

All this can be done by sending a string to the device containing the required
command, for example:

.Dl % printf $ Ns So \ex8b Sc > /dev/tm1637/0

The
.Nm
is intended for those cases when the software for executing these commands is
limited in the use of "bad characters" such as "&", "|", ";" (See
.Xr wordexp 3
.Ar RETURN VALUES )

.Sh SEE ALSO
.Xr tm1637 4 ,
.Xr tm1637d 8 ,
.Xr bh1750 4 ,
.Xr bh1750-daemon 8
.Sh AUTHORS
.An -nosplit
.An Alexander Mishin Aq mishin@mh.net.ru


tm1637ctl.c
/*-
 * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
 *
 * Copyright (C) 2021, Alexander Mishin
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>
#include <stdbool.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <getopt.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <err.h>
#include "tm1637d.h"

static void
usage()
{
    fprintf(stderr, "usage: %s [-t <0|1>] [-l <0..7>] [-d <device>] [-h]\n\n",
	getprogname());
    fprintf(stderr,
    "Optioqns:\n"
	"    -d, --device <device>\n"
	"                        Define a tm1637-device name for control;\n"
	"    -t, --turn <0|1>    Turn a display on or off;\n"
	"    -l, --brightness <0..7>\n"
	"                        Set a brightness level;\n"
	"    -h, --help          Print this help.\n"
    );
}

int
main(int argc, char **argv)
{
    static int dev;
    static char *dev_tm1637 = "/dev/tm1637";

    static int ch, long_index = 0;
    static unsigned long value;
    static uint8_t turn = 255, level = 255;
    extern char *optarg, *end;

    static struct option long_options[] = {
	{"device",     required_argument, 0, 'd' },
	{"turn",       required_argument, 0, 't' },
	{"brightness", required_argument, 0, 'l' },
	{0, 0, 0, 0}
    };

    while ((ch = getopt_long(argc, argv, "t:l:d:h", long_options, &long_index)) != -1) {
	switch(ch) {
	case 'd':
	    dev_tm1637 = optarg;
	    break;
	case 't':
	    value = strtoul(optarg, &end, 0);
	    if (value <= 1)
		turn = value;
	    else
		errx(EXIT_FAILURE,
		    "Only 1 or 0 is allowed to turn the display on or off ");
	    break;
	case 'l':
	    value = strtoul(optarg, &end, 0);
	    if (value <= 7)
		level = value;
	    else
		errx(EXIT_FAILURE,
		     "Brightness level %lu is out of range. Allowed values are 0..7",
		     value);
	    break;
	case 'h':
	    /* FALLTHROUGH */
	default:
	    usage();
	    exit(EXIT_SUCCESS);
	}
    }

    dev = open(dev_tm1637, O_WRONLY | O_DIRECT);
    if (dev < 0)
	errx(EXIT_FAILURE, "opening tm1637 device");

    if (level <= 7)
	ioctl(dev, TM1637IOC_SET_BRIGHTNESS, &level);

    if (turn == 1)
	ioctl(dev, TM1637IOC_ON);
    else if (turn == 0)
	ioctl(dev, TM1637IOC_OFF);

    close(dev);
}
